README.md
![alt tag](https://github.com/PhamBaTho/BTNavigationDropdownMenu/blob/master/Assets/BTNavigationDropdownLogo.png)

[![Pod Version](https://img.shields.io/cocoapods/v/BTNavigationDropdownMenu.svg?style=flat)](http://cocoadocs.org/docsets/BTNavigationDropdownMenu/)
[![License: MIT](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://github.com/PhamBaTho/BTNavigationDropdownMenu/blob/master/LICENSE)
![Language](https://img.shields.io/badge/language-Swift-brightgreen.svg?style=flat)
<!--[![Build Status](https://travis-ci.org/PhamBaTho/BTNavigationDropdownMenu.svg?branch=master)](https://travis-ci.org/PhamBaTho/BTNavigationDropdownMenu)-->

## Introduction
The elegant yet functional dropdown menu, written in Swift, appears underneath the navigation bar to display a list of defined items when a user clicks on the navigation title.

![alt tag](https://github.com/PhamBaTho/BTNavigationDropdownMenu/blob/master/Assets/Demo.gif)

## Installation
### CocoaPods
**BTNavigationDropdownMenu** is available through [CocoaPods](http://cocoapods.org). To install it, simply add the following line to your Podfile:

```ruby
use_frameworks!
pod 'BTNavigationDropdownMenu'
```
Go ahead and **import** BTNavigationDropdownMenu into your own Swift files 
```swift
import BTNavigationDropdownMenu
```

## Carthage
[Carthage](https://github.com/Carthage/Carthage) is a decentralized dependency manager that builds your dependencies and provides you with binary frameworks.

You can install Carthage with [Homebrew](https://brew.sh/) using the following command:

```bash
$ brew update
$ brew install carthage
```

To integrate BTNavigationDropdownMenu into your Xcode project using Carthage, specify it in your `Cartfile`:

```ogdl
github "PhamBaTho/BTNavigationDropdownMenu" ~> 0.7
```

Run `carthage update` to build the framework and drag the built `BTNavigationDropdownMenu.framework` into your Xcode project.

## Usage
### Instantiate
Start by creating an Array that contains strings as **elements of a dropdown list**:
```swift
let items = ["Most Popular", "Latest", "Trending", "Nearest", "Top Picks"]
```
Create a **new instance** of BTNavigationDropdownMenu:
```swift
let menuView = BTNavigationDropdownMenu(navigationController: self.navigationController, containerView: self.navigationController!.view, title: BTTitle.title("Dropdown Menu"), items: items)
```
or like this:
```swift
let menuView = BTNavigationDropdownMenu(title: BTTitle.index(1), items: items)
```

BTTitle is an enum. We can set `BTTitle.title(string)` or `BTTitle.index(Int)` (Note: `BTTitle.index` is the index of defined `items` array).

By default, `navigationController` is the topmost navigation controller and `containerView` is keyWindow. 

(`keyWindow` is recommended for `containerView` so that the black overlay can cover the whole screen. In some cases, `keyWindow` doesn't work properly, like using with side menu (e.g. SWRevealViewController), the dropdown menu doesn't move along with the parent view or navigation controller. To workaround this issue, you can use `self.navigationController!.view`.)

Set **title of navigation bar** as menuView:
```swift
self.navigationItem.titleView = menuView
```
Call BTNavigationDropdownMenu closure to get **the index of selected cell**:
```swift
menuView.didSelectItemAtIndexHandler = {[weak self] (indexPath: Int) -> () in
            print("Did select item at index: \(indexPath)")
            self.selectedCellLabel.text = items[indexPath]
}
```
`menuView.show()` or `menuView.hide()` to show or hide dropdown menu manually.

`menuView.toggle()` to toogle dropdown menu shown/hide.

`menuView.isShown` (Boolean type) property to check showing state of dropdown menu.

`menuView.updateItems(items: [AnyObject])` to update items in dropdown menu.

### Customization
First, assign the items and frame for the dropdown menu. Then, customize the look and feel of the menu by overriding these properties:

| Name | Description | Default value |
|---|---|:---:|
| `navigationBarTitleFont` |  Navigation bar title font | System font, bold, size 17 |
| `cellHeight` | Cell height | 50 |
| `arrowImage` | Arrow next to the navigation title | Provided image |
| `arrowPadding` | Padding between the navigation title and arrow | 15 |
| `cellBackgroundColor` | Cell background color | UIColor.white |
| `cellSeparatorColor` | Cell separator color | UIColor.darkGray |
| `cellTextLabelColor` | Text color inside of the cell | UIColor.darkGray |
| `cellTextLabelFont` | Text font inside the cell | System font, bold, size 17 |
| `cellTextLabelAlignment` | Text alignment inside of the cell | NSTextAlignment.left |
| `cellSelectionColor` | Selected color of the cell | UIColor.lightGray |
| `checkMarkImage` | Checkmark icon for the cell | Provided image |
| `animationDuration` | Animation duration for showing/hiding of the menu | 0.3 |
| `maskBackgroundColor` | Mask layer color | UIColor.black |
| `maskBackgroundOpacity` | Opacity of the mask layer | 0.3 |
| `menuTitleColor` | Title color displayed in the menu | UIColor.lightGray |
| `shouldKeepSelectedCellColor` | Selected cell color visibility when menu is shown | false |
| `shouldChangeTitleText` | Ability to change the selected title text | true |
| `selectedCellTextLabelColor` | Selected cell text label color | UIColor.darkGray |
| `arrowTintColor` | Tint color of the arrow | UIColor.white |

## Requirement
- iOS 9.0+

## Changelog
See the [CHANGELOG](https://github.com/PhamBaTho/BTNavigationDropdownMenu/blob/master/CHANGELOG.md) for details

## License
BTNavigationDropdownMenu is available under the MIT License. See the [LICENSE](https://github.com/PhamBaTho/BTNavigationDropdownMenu/blob/master/LICENSE) for details.


CHANGELOG.md
## [0.7](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.7) (16-11-2020)
### Fixed
- Update deprecated index(of:) to firstIndex(of:) [(#156)](https://github.com/PhamBaTho/BTNavigationDropdownMenu/pull/156)
- Remove ImplicitlyUnwrappedOptional to be compatible with Swift 5.

### Improved
- Update project configuration to be compatible with Swift 5 and Xcode 10.2+
- Use system font as default value of `navigationBarTitleFont`.

## [0.6](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.6) (19-02-2019)
- Merged PR [#145](https://github.com/PhamBaTho/BTNavigationDropdownMenu/pull/145): Carthage support.

## [0.5](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.5) (19-08-2018)
- Support Xcode 10 and Swift 4.2

## [0.4.1](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.4.1) (01-08-2016)
- Merge PR [#71](https://github.com/PhamBaTho/BTNavigationDropdownMenu/pull/71): Now you can configure selectedCellTextLabelColor. arrowTintColor, navigationBarTitleFont
- Merge PR [#75](https://github.com/PhamBaTho/BTNavigationDropdownMenu/pull/75): Now you can keep navigation title static when selecting items in dropdown menu
- Change keepSelectedCellColor to shouldKeepSelectedCellColor
- Fix a bug in updateItems() when updating menu view after HTTP request.

## [0.4](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.4) (23-06-2016)
- Resolve issue [#57](https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/57) and [#63](https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/63). Now you can scroll to the last one when the list is longer than screen's height.
- Resolve issue [#62](https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/62) (Alternate scope initialization). Now you can initialize menuView where `title` parameter doesn't need to be one of the items array.
- Resolve issue [#68](https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/68). Now, you can use `updateItems(items: [AnyObject])` to update items for the dropdown menu after it has been initialized.
- Add containView as parameter when initializing menuView.

## [0.3](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.3) (17-04-2016)
- Add toggle() functions
- Add keepSelectedCellColor property
- Fixed issues [#41](https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/41)
- Fixed issues [#38](https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/38)

## [0.2.1](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.2.1) (10-01-2016)
- Public show() and hide() functions
- Public isShown property

## [0.2](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.2) (15-12-2015)
- Passed navigationController when initializing
- Added the alignment of the text inside cell
- Fixed issue [#16](https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/16): Sometimes navigationController is nil

## [0.1.8](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.8) (15-09-2015)
- Fixed issue: Sometimes navigationController is nil

## [0.1.7](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.7) (04-09-2015)
- Use optional for some properties that relating to navigation controller

## [0.1.6](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.6) (28-08-2015)
- No need to use containerView param when initializing menuView now

## [0.1.5](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.5) (27-08-2015)
- Support device orientation
- Support Split View on iPad

## [0.1.4](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.4) (26-08-2015)
- Fixed issue [#5] (https://github.com/PhamBaTho/BTNavigationDropdownMenu/issues/5). The dropdown overlay appears overtop the tab bar now
- Refactoring and clean up

## [0.1.3](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.3) (25-07-2015)
- Use `usingSpringWithDamping` instead of `contentOffset` for show/hide animation

## [0.1.2](https://github.com/PhamBaTho/BTNavigationDropdownMenu/releases/tag/0.1.2) (21-07-2015)
- Allow change of arrow image after creation of menu


Demo/Demo/ViewController.swift
//
//  ViewController.swift
//  BTNavigationDropdownMenu
//
//  Created by Pham Ba Tho on 6/8/15.
//  Copyright (c) 2015 PHAM BA THO. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var selectedCellLabel: UILabel!
    var menuView: BTNavigationDropdownMenu!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        let items = ["Most Popular", "Latest", "Trending", "Nearest", "Top Picks"]
        self.selectedCellLabel.text = items.first
        self.navigationController?.navigationBar.isTranslucent = false
        self.navigationController?.navigationBar.barTintColor = UIColor(red: 0.0/255.0, green:180/255.0, blue:220/255.0, alpha: 1.0)
        self.navigationController?.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor: UIColor.white]

        // "Old" version
        // menuView = BTNavigationDropdownMenu(navigationController: self.navigationController, containerView: self.navigationController!.view, title: "Dropdown Menu", items: items)
       
        

        menuView = BTNavigationDropdownMenu(navigationController: self.navigationController, containerView: self.navigationController!.view, title: BTTitle.index(2), items: items)

        // Another way to initialize:
        // menuView = BTNavigationDropdownMenu(navigationController: self.navigationController, containerView: self.navigationController!.view, title: BTTitle.title("Dropdown Menu"), items: items)
        
        let maxString = String(repeating: "一", count: 11)
        let maxTitleSize = (maxString as NSString).size(withAttributes: [NSAttributedString.Key.font:menuView.navigationBarTitleFont])
        menuView.menuTitleMaxWidth = maxTitleSize.width
        menuView.menuTitleBreakModel = .byTruncatingMiddle
        menuView.cellHeight = 50
        menuView.cellBackgroundColor = self.navigationController?.navigationBar.barTintColor
        menuView.cellSelectionColor = UIColor(red: 0.0/255.0, green:160.0/255.0, blue:195.0/255.0, alpha: 1.0)
        menuView.shouldKeepSelectedCellColor = true
        menuView.cellTextLabelColor = UIColor.white
        menuView.cellTextLabelFont = UIFont(name: "Avenir-Heavy", size: 17)
        menuView.cellTextLabelAlignment = .left // .Center // .Right // .Left
        menuView.arrowPadding = 15
        menuView.animationDuration = 0.5
        menuView.maskBackgroundColor = UIColor.black
        menuView.maskBackgroundOpacity = 0.3
        menuView.didSelectItemAtIndexHandler = {(indexPath: Int) -> Void in
            print("Did select item at index: \(indexPath)")
            self.selectedCellLabel.text = items[indexPath]
        }
        
        self.navigationItem.titleView = menuView
    }
}
